File: src/calculus_limits_algebraic_medium.py
"""
calculus - limits_algebraic (medium)
Generated: 2026-02-22T04:58:42.983569
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Direct substitution limit (1300-1400)
        a_val = nonzero(-5, 5)
        b_val = nonzero(-8, 8)
        c_val = nonzero(-10, 10)
        
        expr = a_val * x**2 + b_val * x + c_val
        limit_point = nonzero(-4, 4)
        ans = expr.subs(x, limit_point)
        
        return problem(
            question=f"\\lim_{{x \\to {limit_point}}} \\left({latex(expr)}\\right)",
            answer=ans,
            difficulty=(1300, 1400),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Since ${latex(expr)}$ is a polynomial, it is continuous everywhere",
                f"Substitute $x = {limit_point}$ directly",
                f"${latex(expr.subs(x, limit_point))} = {ans}$"
            )
        )
    
    elif problem_type == 2:
        # Factoring to remove indeterminate form (1400-1500)
        root = nonzero(-5, 5)
        other_val = nonzero(-6, 6)
        while other_val == root:
            other_val = nonzero(-6, 6)
        
        numerator = expand((x - root) * (x - other_val))
        denominator = x - root
        
        ans = other_val
        
        return problem(
            question=f"\\lim_{{x \\to {root}}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1400, 1500),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Factor the numerator: ${latex(numerator)} = {latex(factor(numerator))}$",
                f"Cancel common factor $(x - {root})$",
                f"$\\lim_{{x \\to {root}}} (x - {other_val}) = {root} - {other_val} = {ans}$"
            )
        )
    
    elif problem_type == 3:
        # Rationalize conjugate (1500-1600)
        a_val = choice([1, 4, 9, 16, 25])
        b_val = nonzero(-3, 3)
        
        limit_point = a_val
        numerator = sqrt(x) - sqrt(a_val)
        denominator = x - a_val
        
        ans = Rational(1, 2 * sqrt(a_val))
        
        return problem(
            question=f"\\lim_{{x \\to {a_val}}} \\frac{{\\sqrt{{x}} - {latex(sqrt(a_val))}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1500, 1600),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Multiply by conjugate: $\\frac{{\\sqrt{{x}} + {latex(sqrt(a_val))}}}{{\\sqrt{{x}} + {latex(sqrt(a_val))}}}$",
                f"Numerator becomes: $(\\sqrt{{x}} - {latex(sqrt(a_val))})(\\sqrt{{x}} + {latex(sqrt(a_val))}) = x - {a_val}$",
                f"Cancel $(x - {a_val})$ factor",
                f"$\\lim_{{x \\to {a_val}}} \\frac{{1}}{{\\sqrt{{x}} + {latex(sqrt(a_val))}}} = \\frac{{1}}{{{latex(2*sqrt(a_val))}}} = {latex(ans)}$"
            )
        )
    
    elif problem_type == 4:
        # Difference of squares or simple factoring (1400-1500)
        root = nonzero(-4, 4)
        k = nonzero(2, 5)
        
        numerator = expand(k * (x - root) * (x + root))
        denominator = x - root
        
        ans = 2 * k * root
        
        return problem(
            question=f"\\lim_{{x \\to {root}}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1400, 1500),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Factor the numerator: ${latex(numerator)} = {latex(k)}(x - {root})(x + {root})$",
                f"Cancel $(x - {root})$",
                f"$\\lim_{{x \\to {root}}} {latex(k)}(x + {root}) = {latex(k)}({root} + {root}) = {ans}$"
            )
        )
    
    else:
        # Polynomial long division style (1300-1400)
        a_val = nonzero(-3, 3)
        b_val = nonzero(-5, 5)
        
        expr = (a_val * x + b_val) / x
        ans = a_val
        
        return problem(
            question=f"\\lim_{{x \\to \\infty}} \\frac{{{latex(a_val * x + b_val)}}}{{x}}",
            answer=ans,
            difficulty=(1300, 1400),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Divide each term by $x$",
                f"$\\lim_{{x \\to \\infty}} \\left({a_val} + \\frac{{{b_val}}}{{x}}\\right)$",
                f"As $x \\to \\infty$, $\\frac{{{b_val}}}{{x}} \\to 0$",
                f"The limit is ${a_val}$"
            )
        )

emit(generate())

File: src/calculus_limits_algebraic_medium.jl
# calculus - limits_algebraic (medium)
# Generated: 2026-03-08T20:36:37.276769

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/limits_algebraic")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Limit by direct substitution (simplest case)
        q = rand_quadratic(x; a=(-5,5), b=(-10,10), c=(-10,10))
        a = randint(-8, 8)
        ans = substitute(q.expr, x => a)
        
        problem(
            question="Evaluate \\lim_{x \\to $(a)} $(tex(q.expr))",
            answer=ans,
            difficulty=(1200, 1400),
            solution=steps(
                sol("Given", q.expr),
                "The function is continuous at x = $(a), so substitute directly",
                sol("Substituting x = $(a)", ans)
            ),
            time=60
        )
        
    elseif problem_type == 2
        # Limit with factoring (removable discontinuity)
        r = randint(-10, 10)
        a = nonzero(-5, 5)
        b = randint(-12, 12)
        
        numerator = expand(a * (x - r) * (x - b))
        denominator = x - r
        ans = substitute(a * (x - b), x => r)
        
        problem(
            question="Evaluate \\lim_{x \\to $(r)} \\frac{$(tex(numerator))}{$(tex(denominator))}",
            answer=ans,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given", numerator / denominator),
                "Factor numerator: $(tex(a * (x - r) * (x - b)))",
                "Cancel common factor (x - $(r)): $(tex(a * (x - b)))",
                sol("Substitute x = $(r)", ans)
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Limit with rationalization (conjugate)
        a = randint(1, 20)
        h_sym = choice([:h, :x])
        @variables h
        var = h_sym == :h ? h : x
        
        numerator = sqrt(a + var) - sqrt(a)
        ans = 1 // (2 * sqrt(a))
        
        limit_expr = h_sym == :h ? "h \\to 0" : "x \\to 0"
        
        problem(
            question="Evaluate \\lim_{$(limit_expr)} \\frac{$(tex(numerator))}{$(tex(var))}",
            answer=ans,
            difficulty=(1500, 1700),
            solution=steps(
                "Multiply by conjugate: \\frac{\\sqrt{$(a) + $(h_sym)} + \\sqrt{$(a)}}{\\sqrt{$(a) + $(h_sym)} + \\sqrt{$(a)}}",
                "Numerator becomes: $(a) + $(h_sym) - $(a) = $(h_sym)",
                "Simplify: \\frac{1}{\\sqrt{$(a) + $(h_sym)} + \\sqrt{$(a)}}",
                sol("Substitute $(h_sym) = 0", ans)
            ),
            time=150
        )
        
    elseif problem_type == 4
        # Limit with polynomial division
        r1 = randint(-8, 8)
        r2 = randint(-8, 8)
        while r2 == r1
            r2 = randint(-8, 8)
        end
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        
        numerator = expand(a * (x - r1) * (x - r2))
        denominator = expand(b * (x - r1))
        ans = substitute(a * (x - r2) // b, x => r1)
        
        problem(
            question="Evaluate \\lim_{x \\to $(r1)} \\frac{$(tex(numerator))}{$(tex(denominator))}",
            answer=ans,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given", numerator / denominator),
                "Factor numerator: $(tex(a * (x - r1) * (x - r2)))",
                "Factor denominator: $(tex(b * (x - r1)))",
                "Cancel (x - $(r1)): \\frac{$(tex(a * (x - r2)))}{$(b)}",
                sol("Substitute x = $(r1)", ans)
            ),
            time=120
        )
        
    else
        # Limit with complex rational expression
        r = randint(-6, 6)
        a = nonzero(-6, 6)
        b = randint(-10, 10)
        c = nonzero(-6, 6)
        d = randint(-10, 10)
        
        numerator = expand(a * (x - r) * (x - b))
        denominator = expand(c * (x - r) * (x - d))
        
        ans = substitute((a * (x - b)) / (c * (x - d)), x => r)
        
        problem(
            question="Evaluate \\lim_{x \\to $(r)} \\frac{$(tex(numerator))}{$(tex(denominator))}",
            answer=ans,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Given", numerator / denominator),
                "Factor numerator: $(tex(a * (x - r) * (x - b)))",
                "Factor denominator: $(tex(c * (x - r) * (x - d)))",
                "Cancel common factor (x - $(r)): \\frac{$(tex(a * (x - b)))}{$(tex(c * (x - d)))}",
                sol("Substitute x = $(r)", ans)
            ),
            time=150
        )
    end
end